From aeec6ad4a7c5ad9fb8700c0486d7f6650e0ee540 Mon Sep 17 00:00:00 2001
From: Andrew Liu <shengping.liu@windriver.com>
Date: Tue, 27 Oct 2009 17:26:40 +0800
Subject: [PATCH 3/3] PCIE: support one PCIE slot on fsl_8572ds board

As of this time, just one PCIE-slot is supported
which is the first slot from the top down inside
the board case.

Signed-off-by: Andrew Liu <shengping.liu@windriver.com>
---
 arch/powerpc/include/asm/paravirt.h |    4 +++
 arch/powerpc/include/asm/wrhv.h     |    3 ++
 arch/powerpc/kernel/paravirt.c      |   11 +++++++++
 arch/powerpc/kernel/pci-common.c    |    9 +++++++-
 arch/powerpc/kernel/vbi/wrhv.c      |   39 +++++++++++++++++++++++++++++++++++
 5 files changed, 65 insertions(+), 1 deletions(-)

diff --git a/arch/powerpc/include/asm/paravirt.h b/arch/powerpc/include/asm/paravirt.h
index 0af22f7..826ce48 100644
--- a/arch/powerpc/include/asm/paravirt.h
+++ b/arch/powerpc/include/asm/paravirt.h
@@ -55,6 +55,10 @@ struct pv_cpu_ops {
                                struct pt_regs *linux_regs);
 };
 
+#ifdef CONFIG_PCI
+extern int native_pci_read_irq_line(struct pci_dev *pci_dev);
+#endif
+
 /* general info */
 struct pv_info {
         const char *name;
diff --git a/arch/powerpc/include/asm/wrhv.h b/arch/powerpc/include/asm/wrhv.h
index 84a4881..48a8ab2 100644
--- a/arch/powerpc/include/asm/wrhv.h
+++ b/arch/powerpc/include/asm/wrhv.h
@@ -24,6 +24,9 @@ extern unsigned int wrhv_vioapic_get_irq(void);
 extern void wrhv_init_irq(void);
 extern void __init wrhv_calibrate_decr(void);
 extern void __init wrhv_time_init(void);
+#ifdef CONFIG_PCI
+extern int fsl8572_get_pci_intr_wrhv(struct pci_dev *dev);
+#endif
 
 extern unsigned long wrhv_cpu_freq;
 
diff --git a/arch/powerpc/kernel/paravirt.c b/arch/powerpc/kernel/paravirt.c
index eb01d5f..e8e57ba 100644
--- a/arch/powerpc/kernel/paravirt.c
+++ b/arch/powerpc/kernel/paravirt.c
@@ -83,6 +83,7 @@
 #include <asm/current.h>
 #include <asm/processor.h>
 
+#include <asm/wrhv.h>
 #include <asm/paravirt.h>
 
 #ifdef CONFIG_WRHV
@@ -253,6 +254,16 @@ inline int paravirt_enabled(void)
         return pv_info.paravirt_enabled;
 }
 
+#ifdef CONFIG_PCI
+int paravirt_pci_read_irq_line(struct pci_dev *dev)
+{
+#ifdef CONFIG_WRHV_8572
+	return fsl8572_get_pci_intr_wrhv(dev);
+#endif /* CONFIG_WRHV_8572 */
+	return -1;
+}
+#endif
+
 extern struct pv_time_ops pv_time_ops;
 extern struct pv_cpu_ops pv_cpu_ops;
 extern struct pv_irq_ops pv_irq_ops;
diff --git a/arch/powerpc/kernel/pci-common.c b/arch/powerpc/kernel/pci-common.c
index ea0c61e..1ce5415 100644
--- a/arch/powerpc/kernel/pci-common.c
+++ b/arch/powerpc/kernel/pci-common.c
@@ -180,12 +180,14 @@ char __devinit *pcibios_setup(char *str)
 	return str;
 }
 
+int paravirt_pci_read_irq_line(struct pci_dev *pci_dev)
+        __attribute__((weak, alias("native_pci_read_irq_line")));
 /*
  * Reads the interrupt pin to determine if interrupt is use by card.
  * If the interrupt is used, then gets the interrupt line from the
  * openfirmware and sets it in the pci_dev and pci_config line.
  */
-int pci_read_irq_line(struct pci_dev *pci_dev)
+int native_pci_read_irq_line(struct pci_dev *pci_dev)
 {
 	struct of_irq oirq;
 	unsigned int virq;
@@ -253,6 +255,11 @@ int pci_read_irq_line(struct pci_dev *pci_dev)
 
 	return 0;
 }
+
+int pci_read_irq_line(struct pci_dev *pci_dev)
+{
+	return paravirt_pci_read_irq_line(pci_dev);
+}
 EXPORT_SYMBOL(pci_read_irq_line);
 
 /*
diff --git a/arch/powerpc/kernel/vbi/wrhv.c b/arch/powerpc/kernel/vbi/wrhv.c
index 17f61d8..b1d2e04 100644
--- a/arch/powerpc/kernel/vbi/wrhv.c
+++ b/arch/powerpc/kernel/vbi/wrhv.c
@@ -996,3 +996,42 @@ void wrhv_init(void)
 		wrhv_early_init_dt_scan_memory;
 
 }
+
+#if defined(CONFIG_WRHV_8572) && defined(CONFIG_PCI)
+int fsl8572_get_pci_intr_wrhv(struct pci_dev *dev)
+{
+	int irq;
+	u8  pin;
+	char VectorName[8];
+	
+	pci_read_config_byte(dev, PCI_INTERRUPT_PIN, &pin);
+
+	switch (dev->bus->number) {
+	case 0x01:
+		switch (pin) {
+		case 0x01:
+			sprintf(VectorName, "%s", "PCIe2_A");
+			break;
+		case 0x02:
+			sprintf(VectorName, "%s", "PCIe2_B");
+			break;
+		case 0x03:
+			sprintf(VectorName, "%s", "PCIe2_C");
+			break;
+		case 0x04:
+			sprintf(VectorName, "%s", "PCIe2_D");
+			break;
+		}
+		irq = vbiIntVecFind(VectorName, 1);
+		break;
+	default:
+		break;
+	}
+	
+	dev->irq = irq;
+        printk("WRHV-PCI: BUS NO:%x CLASS:%x IRQ%d\n",
+                dev->bus->number, dev->class, dev->irq);
+	
+	return irq;
+}
+#endif /* CONFIG_WRHV_8572 & CONFIG_PCI*/
-- 
1.6.5.2

